commons/src/main/java/io/aiven/kafka/tieredstorage/commons/storage/s3/S3Storage.java

```diff
@@ -0,0 +1,103 @@
+import io.aiven.kafka.tieredstorage.commons.storage.FileUploader;
+import io.aiven.kafka.tieredstorage.commons.storage.InvalidRangeException;
+import io.aiven.kafka.tieredstorage.commons.storage.KeyNotFoundException;
+import io.aiven.kafka.tieredstorage.commons.storage.StorageBackEndException;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.AmazonS3Exception;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+import com.amazonaws.services.s3.model.S3Object;
+
+public class S3Storage implements FileUploader, FileDeleter, FileFetcher {
+
+    final AmazonS3 s3Client;
+    final String bucketName;
+
+    public S3Storage(final AmazonS3 s3Client, final String bucketName) {
+        this.s3Client = s3Client;
+        this.bucketName = bucketName;
+    }
+
+    @Override
+    public void upload(final InputStream inputStream, final String key) throws StorageBackEndException {
+        try (final S3MultiPartOutputStream out = new S3MultiPartOutputStream(bucketName, key, s3Client)) {
+            inputStream.transferTo(out);
+        } catch (final AmazonS3Exception | IOException e) {
+            throw new StorageBackEndException("Failed to upload " + key, e);
+        }
+    }
+
+    @Override
+    public void delete(final String key) throws StorageBackEndException {
+        try {
+            s3Client.deleteObject(bucketName, key);
+        } catch (final AmazonS3Exception e) {
+            throw new StorageBackEndException("Failed to delete " + key, e);
+        }
+    }
+
+    @Override
+    public InputStream fetch(final String key) throws StorageBackEndException {
+        try {
+            final GetObjectRequest getRequest = new GetObjectRequest(bucketName, key);
+            final S3Object object = s3Client.getObject(getRequest);
+            return object.getObjectContent();
+        } catch (final AmazonS3Exception e) {
+            if (e.getStatusCode() == 404) {
+                throw new KeyNotFoundException(this, key, e);
+            } else {
+                throw new StorageBackEndException("Failed to fetch " + key, e);
+            }
+        }
+    }
+
+    @Override
+    public InputStream fetch(final String key, final BytesRange range) throws StorageBackEndException {
+        try {
+            final GetObjectRequest getRequest = new GetObjectRequest(bucketName, key);
+            getRequest.setRange(range.from, range.to);
+            final S3Object object = s3Client.getObject(getRequest);
+            return object.getObjectContent();
+        } catch (final AmazonS3Exception e) {
+            if (e.getStatusCode() == 404) {
+                throw new KeyNotFoundException(this, key, e);
+            }
+            if (e.getStatusCode() == 416) {
+                throw new InvalidRangeException("Invalid range " + range, e);
+            }
+
+            throw new StorageBackEndException("Failed to fetch " + key, e);
+        }
+    }
```

Comment:
These two share most of the code, probably we can have a common private function that works on `GetObjectRequest`? But that's just a suggestion, your call.

Reply:
Yeah, I guess we can leave it as it is for the time being and if the request gets more complex and grow together we can refactor.

Reply:
sure